Week4/HW04.hs

```diff
@@ -0,0 +1,149 @@
+{-# OPTIONS_GHC -Wall #-}
+
+-- https://www.cis.upenn.edu/~cis194/spring15/hw/04-typeclasses.pdf
+
+module Week4.HW04 where
+
+import Data.List (dropWhileEnd)
+newtype Poly a = P [a]
+
+-- Exercise 1 -----------------------------------------
+
+x :: Num a => Poly a
+x = P [0, 1]
+
+-- Exercise 2 ----------------------------------------
+
+instance (Num a, Eq a)              => Eq (Poly a) where
+    -- This will not be effecient way of doing it I think,
+    -- because its taking extra storage to save the new list 
+    -- without trailing zeros and then does the matching
+    P a == P b                      = withoutTrailingZerosA == withoutTrailingZerosB
+        where 
+            withoutTrailingZerosA   = dropWhileEnd (== 0) a 
+            withoutTrailingZerosB   =  dropWhileEnd (== 0) b
+        
+
+-- Exercise 3 -----------------------------------------
+
+-- e == exponent 
+-- c == count of the polynomial term, is it first, second, third and so on
+showHelper :: (Num a, Show a, Eq a) => Poly a -> Int -> Int -> String
+showHelper (P (p:ps)) e c
+    | p == 0                = showHelper (P ps) (e-1) (c+1)
+    | e == 0                = leadingPlus ++ show p 
+    -- I was trying to get nested guards here instead of using [if-then-else] or [case] but wasnt able to 
```

Comment:
Not sure what you mean by nested guards. Can you give your desired goal syntax that you'd like to achieve?

Reply:
I wanted something like this (ignore the silly example)
There was some multi-way if warning 
Looked it up
had to add this to top of file and the code works then 
> {-# LANGUAGE MultiWayIf #-}

But is it ok to do things this way or refrain from using it and go with case in such cases ?

<img width="451" alt="Screenshot 2021-01-26 at 1 52 45 AM" src="https://user-images.githubusercontent.com/30658517/105761455-316f6e80-5f79-11eb-8a55-1dc77044652a.png">

Reply:
If you just need to compare for equality, than case (nested) is just fine.
Because in the end even if-s and multiway if-s are in the end desugared to case statements (if interested, I recommend this video which shows how most of haskell surface syntax is desugared to just a couple of simple constructs: https://www.youtube.com/watch?v=fty9QL4aSRc).

In general I don't worry about using syntactic extensions and I use them freely whenever I feel like it improves code readability. And unlike more involved type-level language extensions, you can always easily refactor the code to not-using-them.

Reply:
got it, thanks :v: